plugins/modules/packaging/language/ansible_galaxy_install.py

```diff
@@ -0,0 +1,318 @@
+                    continue
+                if self.vars.dest is not None and match.group('path') != self.vars.dest:
+                    current_path = None
+                    continue
+                current_path = match.group('path') if match else None
+                elems_dict[current_path] = {}
+
+            elif current_path is not None:
+                match = elem_re.match(line)
+                if not match or (self.vars.name is not None and match.group('elem') != self.vars.name):
+                    continue
+                elems_dict[current_path][match.group('elem')] = match.group('version')
+        return elems_dict
+
+    def __list_collections__(self):
+        return self.__list_element__('collection', self._RE_LIST_PATH, self._RE_LIST_COLL)
+
+    def __list_roles__(self):
+        return self.__list_element__('role', self._RE_LIST_PATH, self._RE_LIST_ROLE)
+
+    def __setup29__(self):
+        self.vars.set("new_collections", {})
+        self.vars.set("new_roles", {})
+        self.vars.set("ansible29_change", False, change=True, output=False)
+        if not self.vars.ack_ansible29:
+            self.module.warn("Ansible 2.9 or older: unable to retrieve lists of roles and collections already installed")
+            if self.vars.requirements_file is not None and self.vars.type == 'both':
+                self.module.warn("Ansible 2.9 or older: will install only roles from requirement files")
+
+    def __setup210plus__(self):
+        self.vars.set("new_collections", {}, change=True)
+        self.vars.set("new_roles", {}, change=True)
+        if self.vars.type != "collection":
+            self.vars.installed_roles = self.__list_roles__()
+        if self.vars.type != "roles":
+            self.vars.installed_collections = self.__list_collections__()
+
+    def __run__(self):
+        if self.is_ansible29:
+            if self.vars.type == 'both':
+                raise ValueError("Type 'both' not supported in Ansible 2.9")
+            self.__setup29__()
+        else:
+            self.__setup210plus__()
+        params = ('type', {'galaxy_cmd': 'install'}, 'force', 'dest', 'requirements_file', 'name')
+        self.run_command(params=params)
+
+    def process_command_output(self, rc, out, err):
+        for line in out.splitlines():
+            match = self._RE_INSTALL_OUTPUT.match(line)
+            if not match:
+                continue
+            if match.group("collection"):
+                self.vars.new_collections[match.group("collection")] = match.group("cversion")
+                if self.is_ansible29:
+                    self.vars.ansible29_change = True
+            elif match.group("role"):
+                self.vars.new_roles[match.group("role")] = match.group("rversion")
+                if self.is_ansible29:
+                    self.vars.ansible29_change = True
+
+
+def main():
+    galaxy = AnsibleGalaxyInstall()
+    galaxy.run()
+
+
+if __name__ == '__main__':
+    main()
```

Comment:
I haven't been watching this repo for quite a while and from what I see, there's an attempt to generalize structuring the module implementations.

Since I've experimented with this a while ago, I thought I'd drop an idea here.

Basically, having a two-step init+run pattern repeated everywhere is rather fragile. It looks like a one-step thing on the top level. It's easy to solve this by having a class method that wraps the operations. This will also allow you to avoid defining an intermediary `main()` function for as long as it only does just this one operation.

The the patch would look like this:
```suggestion

    @classmethod
    def execute(cls):
        cls().run()


if __name__ == '__main__':
    AnsibleGalaxyInstall.execute()
```

Of course, this is a bit out of the scope here but I believe it'd be really useful if added to the base class in a separate PR (optionally, the classmethod could accept args+kwargs that would be bypassed to the initializer).

Here's the module I structured like this, maybe you'll pick up some other ideas: https://github.com/webknjaz/ansible-gentoo-laptop/blob/master/library/eix.py.

Reply:
Well, a lot to unpack here.

> I haven't been watching this repo for quite a while and from what I see, there's an attempt to generalize structuring the module implementations.

There is on my part one such attempt, that was achieved by creating this set of classes and mixins under module_utils: https://github.com/ansible-collections/community.general/tree/main/plugins/module_utils/mh

It provides the same features I've seen in the EIX example you sent, plus other things as well. I consider it alpha quality yet, because there are quite a number of improvements that could be made in there. Not to mention the lack of documentation as of now.

> Basically, having a two-step init+run pattern repeated everywhere is rather fragile. It looks like a one-step thing on the top level. It's easy to solve this by having a class method that wraps the operations. This will also allow you to avoid defining an intermediary `main()` function for as long as it only does just this one operation.

Sorry, I fail to see how that would be fragile. Could you please expand on that? Also, I try and avoid class methods (and static ones) whenever I can, they look lke we're trying to squeeze old structured programming mentality into OO languages. That said, I have used both on occasions but not without a compelling reason, which I fail to find here.

The use of the intermediary `main()` function is helpful when writing unit tests - in the long run I would like to generalize part of the testing as well, so having a common entry point in all modules is useful for that.

> Of course, this is a bit out of the scope here but I believe it'd be really useful if added to the base class in a separate PR (optionally, the classmethod could accept args+kwargs that would be bypassed to the initializer).

That sounds exactly like `ModuleHelper`.

> Here's the module I structured like this, maybe you'll pick up some other ideas: https://github.com/webknjaz/ansible-gentoo-laptop/blob/master/library/eix.py.

Most of what I see in there is already part of MH: dispatching, support for running commands (in fact your `run_or_raise()` method is overkill - `module.run_command()` has a bool parameter `check_rc` that, when set to `True` makes it pop up an exception when `rc != 0`). I could make better use of Exceptions like you did, but that is also on my TO-DO list.

And finally, it is also my goal with MH is to have a generalized mechanism for modules that interact with:
- HTTP APIs (REST or otherwise)
- third-party libraries (that encapsulate domain-specific actions)

But that is also in my mental backlog for it.

Sorry for the response's length, hope it didn't sound harsh. New ideas, opposing ideas, plain criticism is welcome :-)

Cheers

Reply:
@webknjaz the other modules using MH, for comparison:
https://github.com/ansible-collections/community.general/blob/main/plugins/modules/system/xfconf.py
https://github.com/ansible-collections/community.general/blob/main/plugins/modules/packaging/language/cpanm.py
https://github.com/ansible-collections/community.general/blob/main/plugins/modules/packaging/os/snap.py
https://github.com/ansible-collections/community.general/blob/main/plugins/modules/web_infrastructure/jira.py

Reply:
Most of it makes sense to me but having a class method is not such a wild idea. It's main use case is to provide alternative constructors (like making one object from another). But I still think it's useful to have such an entrypoint in the MH class. In all the examples you've shown I see the same copy-paste init+run. It's an antipattern to force the caller to perform two actions instead of one and then abstract that with main anyway. You could still easily do `main = AnsibleGalaxyInstall.execute` if you want to have a callable named `main()` if you really need to have it (which I don't believe you do — if it's for discovery during testing, you could identify a class that inherits MH dynamically and call `execute()` there, for example).

Why is the two-step pattern harmful? It's not only because of the DRY considerations. It adds mental overhead because one must remember to do a sequence of operations and not to have anything in between. It's encouraging bad behavior — having a mutable instance around means that over time someone may start refactoring the thing, adding instructions in between those two lines. At some point it won't be obvious that they are related.
It could turn from
```
inst = Class()
inst.run()
```
into
```
inst = Class()
call()
call()
call()
call()
call()
# x100
call()
call()
inst.run()
```
```
inst = Class()
call()
call()
call()
setattr(inst, '_magical_thing', X)
call()
call()
# x100
call()
inst._configure_something_weird()
call()
inst.run()
```

Did you design it to be decoupled like this? If not, you'd need a custom linter to enforce that two-line boilerplate or just have a one-step class method as I suggested.

Reply:
Well, that is boilerplate code, no doubt, and it comes straight from Ansible documentation itself:
https://docs.ansible.com/ansible/latest/dev_guide/developing_modules_general.html#creating-a-module

So, whilst I do not disagree with you, I feel like MH is chopping down a larger number of and more importantly more relevant lines of boilerplate code. I think in terms of mental overhead, classes based on MH are way lighter than the ones that are not.

As of the dynamic discovery of the class in the test - wouldn't *that* going to be boilerplate code, only elsewhere? Not to mention that, in the future, we might have more than one MH-based class per file. See the case of: https://github.com/ansible-collections/community.general/pull/2946 - I have been brewing this idea of having multiple implementations and dynamically choosing the one that makes sense for the runtime environment.

Decoupling design: I did not go as far as designing *this* aspect. TBH I did no think this part very thoroughly, because... it wasn't - and still isn't IMO - really all that relevant. I think we have bigger boilerplate-fishes to fry.

I would say this would be a lower priority for now, and clearly not something to be addressed in this PR. 

That said, if you want to pursue that change, please do open an issue, hook it up with this PR for reference and we take it from there. Would that work for you?

Reply:
@webknjaz in fact, I tried removing the `main()` function and calling the MH module directly:
```
ERROR: plugins/modules/system/xfconf_info.py:0:0: missing-main-call: Did not find a call to main()
ERROR: plugins/modules/system/xfconf_info.py:191:0: missing-if-name-main: Next to last line should be: if __name__ == "__main__":
```
That tripped the validation checks.

Reply:
The top-level object being a "command" object which delegates to an instance of `AnsibleModule` is also an incomplete abstraction from my experience. Sometimes that top-level object is a factory which returns a command object. That pattern tends to pop up when we have to migrate/deprecate the supporting backend be it a binary, REST client, third-party library, etc... Keeping the primary execution in `main` is also a more natural evolution from `Ansible-core` where some of these deeper structural questions should really be addressed. (Other collections have their own abstractions which should be considered as well when generalizing)

Reply:
@webknjaz @Ajpantuso guys, I appreciate the discussion, but it seems liek this should be dealt with in another issue or PR. So, unless you believe there is something that must be changed for this PR, I am considering this conversation resolved, and will move on to approvals and merging.